src/Cms/System/UpdateStatus.php

```diff
@@ -0,0 +1,680 @@
+		$filters = [
+			'kirby'  => $this->app->version(),
+			'php'    => phpversion()
+		];
+
+		if ($type === 'plugin') {
+			$filters['plugin'] = $this->currentVersion;
+		}
+
+		$messages = static::filterArrayByVersion(
+			$this->data['messages'] ?? [],
+			$filters
+		);
+
+		// add a message for each vulnerability
+		// the current version is affected by
+		foreach ($this->vulnerabilities() as $vulnerability) {
+			if ($type === 'plugin') {
+				$vulnerability['plugin'] = $this->pluginName;
+			}
+
+			$messages[] = [
+				'text' => I18n::template(
+					'system.issues.vulnerability.' . $type,
+					null,
+					$vulnerability
+				),
+				'link' => $vulnerability['link'] ?? null,
+				'icon' => 'bug'
+			];
+		}
+
+		// add special message for end-of-life versions
+		$versionEntry = $this->versionEntry();
+		if (($versionEntry['status'] ?? null) === 'end-of-life') {
+			$messages[] = [
+				'text' => match ($type) {
+					'plugin' => I18n::template(
+						'system.issues.eol.plugin',
+						null,
+						['plugin' => $this->package->name()]
+					),
+					default => I18n::translate('system.issues.eol.kirby')
+				},
+				'link' => $versionEntry['status-link'] ?? 'https://getkirby.com/security/end-of-life',
+				'icon' => 'bell'
+			];
+		}
+
+		return $this->messages = $messages;
+	}
+
+	/**
+	 * Returns the raw status value
+	 *
+	 * @return string 'up-to-date'|'not-vulnerable'|'security-update'|
+	 *                'security-upgrade'|'update'|'upgrade'|'unreleased'|'error'
+	 */
+	public function status(): string
+	{
+		return $this->targetData()['status'];
+	}
+
+	/**
+	 * Version that is suggested for the update/upgrade
+	 */
+	public function targetVersion(): string|null
+	{
+		return $this->targetData()['version'];
+	}
+
+	/**
+	 * Returns the Panel theme for the status value
+	 *
+	 * @return string 'positive'|'negative'|'info'|'notice'
+	 */
+	public function theme(): string
+	{
+		return match ($this->status()) {
+			'up-to-date', 'not-vulnerable' => 'positive',
+			'security-update', 'security-upgrade' => 'negative',
+			'update', 'upgrade' => 'info',
+			default => 'notice'
+		};
+	}
+
+	/**
+	 * Returns the most important human-readable
+	 * status information as array
+	 */
+	public function toArray(): array
+	{
+		return [
+			'label'   => $this->label(),
+			'theme'   => $this->theme(),
+			'url'     => $this->url(),
+			'version' => $this->currentVersion ?? '–',
+		];
+	}
+
+	/**
+	 * URL of the target version with fallback
+	 * to the URL of the current version;
+	 * `null` is returned if no URL is known
+	 */
+	public function url(): string|null
+	{
+		return $this->targetData()['url'];
+	}
+
+	/**
+	 * Returns all vulnerabilities the current version
+	 * is affected by unless no data is available
+	 */
+	public function vulnerabilities(): array|null
+	{
+		if (isset($this->vulnerabilities) === true) {
+			return $this->vulnerabilities;
+		}
+
+		if (!$this->data || !$this->currentVersion) {
```

Comment:
```suggestion
		if ($this->data === null || $this->currentVersion === null) {
```

Let's be explicit in the if statements, the `!` is always hard to understand what it really is supposed to test for.

Reply:
To be honest my goal with these checks was to be as lax as possible explicitly. If the value ends up to be falsy but not null for some reason, we still need to return early. Checking strictly for null would be a downgrade from that safety. We do have prop and return types everywhere, but this provides an extra level of safety.

I can still change it, but I think it's another one of these general decisions we need to make as a team.

Reply:
IMO the possible types don't really matter here. It's just "if this value is not set, then we cannot continue and need to return early".